src/model_repository_manager.h

```diff
@@ -117,29 +197,111 @@ class ModelRepositoryManager {
+    // i.e. it is safe to move forward without waiting until notified.
+    std::shared_ptr<std::condition_variable> retry_notify_cv_;
   };
 
   // Interface for dependency graph operations
   class DependencyGraph {
    public:
-    DependencyGraph() = delete;
+    // Create a placeholder object to be assigned later. Do not use the
+    // placeholder object before it is assigned.
+    DependencyGraph() : global_map_ptr_(nullptr) {}
 
-    // Passing reference of 'global_map'.
+    // Passing pointer of 'global_map'.
     // There is coupling between the dependency graph and global map: global map
     // will be used to resolve dependency (node connectivity), and global map
     // will be updated to reflect node changes.
     DependencyGraph(
-        std::unordered_map<std::string, std::set<ModelIdentifier>>& global_map)
-        : global_map_ref_(global_map)
+        std::unordered_map<std::string, std::set<ModelIdentifier>>* global_map)
+        : global_map_ptr_(global_map)
     {
     }
 
+    DependencyGraph(const DependencyGraph&) = delete;
+
+    // Copy from rhs, but set global_map_ptr_ to the provided global_map.
+    DependencyGraph(
+        const DependencyGraph& rhs,
+        std::unordered_map<std::string, std::set<ModelIdentifier>>* global_map);
+
+    DependencyGraph& operator=(const DependencyGraph&) = delete;
+
+    // Discard all state of this object, and copy state from rhs, but set
+    // global_map_ptr_ to the provided global_map.
+    void Assign(
+        const DependencyGraph& rhs,
+        std::unordered_map<std::string, std::set<ModelIdentifier>>* global_map);
+
+    void Swap(DependencyGraph& rhs);
+
     std::unordered_map<ModelIdentifier, std::unique_ptr<DependencyNode>>*
     MutableNodes()
     {
       return &nodes_;
     }
 
+    // Look up node in dependency graph with matching model identifier. If not
+    // found and fuzzy match is allowed, a node in different namespace will be
+    // returned if it is the only node with the same name
+    DependencyNode* FindNode(
+        const ModelIdentifier& model_id, const bool allow_fuzzy_matching) const;
+
+    // Set the nodes to lock state. If a node is already locked, then the
+    // identifier of the node is returned. Otherwise, nullptr is returned.
+    // The previous dependency graph is for checking if the node is already
+    // locked when it is not found on this dependency graph, as the node could
+    // be deleted on this graph from the previous, and correctly return the
+    // identifier.
+    // The conflict notifier can be provided optionally. If a node is already
+    // locked, it will be set to a notifier that the calling thread can wait on
+    // until notified to retry, to reduce the number of retries.
+    std::unique_ptr<ModelIdentifier> LockNodes(
+        const std::set<ModelIdentifier>& nodes,
+        const DependencyGraph& prev_dependency_graph,
+        std::shared_ptr<std::condition_variable>* retry_notify_cv = nullptr);
```

Comment:
@GuanLuo what do you think?

Reply:
Similar argument on the `nullptr` in `LoadUnloadModels`, and if not returning model identifier, the signature can even be
`std::shared_ptr<std::condition_variable>LockNodes(...)`, return value == nullptr if nothing to be waited on and contain an cv if there is conflict, and caller can decide if to wait for it or abort.

Reply:
Yes, the only thing is the caller might want to know the `ModelIdentifier` for log printing purpose. Maybe it would be easier to simply return the `DependencyNode*` which contains everything and can be set to `nullptr` if no conflict.